core/array.h

```diff
@@ -98,6 +99,9 @@ class Array {
 	Array duplicate(bool p_deep = false) const;
 
 	Array slice(int p_begin, int p_end, int p_step = 1, bool p_deep = false) const;
+	Array filter(const Callable &p_callable) const;
+	Array map(const Callable &p_callable) const;
+	Variant reduce(const Callable &p_callable, const Variant &p_accum) const;
```

Comment:
I think the `p_accum` can be optional. This makes the algorithm a bit more complicated, since it needs to use the first element of the array as the initial value (so it needs to check the size first). But it's more compatible with most implementations of `reduce`, in which this is an optional value.